src/passes/OptimizeInstructions.cpp

```diff
@@ -1391,6 +1391,19 @@ struct OptimizeInstructions
       right->type = Type::i32;
       return right;
     }
+    {
+      Expression* x;
+      // ~(1 << x) aka (1 << x) ^ -1  ==>  rotl(-2, x)
+      if (matches(curr,
+                  binary(Abstract::Xor,
+                         binary(&curr, Abstract::Shl, ival(1), any(&x)),
```

Comment:
```suggestion
                         binary(Abstract::Shl, ival(1), any(&x)),
```

Reply:
I can't do this due to `curr` reassigned (reused) and I still need access to `curr->left`

Reply:
You can get rid of `&curr` here and still access `curr->left` because `curr` is a `Binary*`.

Reply:
In this case I should do `curr->left->left` (and this not `Binary*`) due to we override (reuse) `curr`.

Reply:
This rule becomes more tricky and unclear in my opinion with all this re-usages.

Reply:
I think this will be much more obvious:
```cpp
{
   // ~(1 << x) aka (1 << x) ^ -1  ==>  rotl(-2, x)
   Binary* inner;
   if (matches(curr,
               binary(Abstract::Xor,
                      binary(&inner, Abstract::Shl, ival(1), any()),
                      ival(-1)))) {
      inner->op = Abstract::getBinary(type, Abstract::RotL);
      inner->left->cast<Const>()->value = Literal::makeFromInt32(-2, type);
      return inner;
   }
}
```

Reply:
Ok, that sounds good to me. Sorry for the confusion!

Reply:
Although this works, too, and is what I had in mind.

```
    {
      // ~(1 << x) aka (1 << x) ^ -1  ==>  rotl(-2, x)
      Expression* x;
      if (matches(curr,
                  binary(Abstract::Xor,
                         binary(Abstract::Shl, ival(1), any(&x)),
                         ival(-1)))) {
        curr->op = Abstract::getBinary(type, Abstract::RotL);
        right->value = Literal::makeFromInt32(-2, type);
        curr->left = right;
        curr->right = x;
      }
    }
```